### DISCUSSION TITLE
Getting env variables inside a plugin

### DISCUSSION THREAD

**Discussion by andreas83** [nuxt/nuxt#18063] [Questions] [answered]
I need to find a way to read env variables inside a client plugin:

nuxt3
plugins/ws.client.js

```
import { defineNuxtPlugin } from '#app'
import dotenv from 'dotenv';

dotenv.config();

export default defineNuxtPlugin(() => {
  return {
    provide: {
      ws: () => new WebSocket(process.env.wsURL)
    }
  }
})
```

Error:

_Uncaught ReferenceError: process is not defined
    at Object._S [as config] (entry-28abbd49.mjs:26)
    at entry-28abbd49.mjs:26_

**Any hints are highly appreciated.**

**Comment by lucpotage** [comment, marked as answer]
Hi! You don't need to install `dotenv`. Nuxt handles this for you.

In your `nuxt.config.ts` file:
```ts
export default defineNuxtConfig({
  publicRuntimeConfig: {
    wsURL: process.env.wsURL
  }
})
```

In your `plugins/ws.client.js` file:
```js
export default defineNuxtPlugin(({ $config }) => {
  return {
    provide: {
      ws: () => new WebSocket($config.wsURL)
    }
  }
})
```

A working example: https://stackblitz.com/edit/github-kgmgvk?file=plugins%2Fws.client.js

**Reply by andreas83** [reply]
Je vous remercie (thank you) !!!

**Reply by notflip** [reply]
@lucpotage How about when that value is a secret? Thanks!

**Reply by lucpotage** [reply]
In this case, new WebSocket runs in the client, so the secret would be shared.

You can learn more about the new runtime config API here: https://nuxt.com/docs/guide/going-further/runtime-config
